src/PIL/DdsImagePlugin.py

```diff
@@ -148,15 +147,38 @@ def _open(self):
             self.tile = [("raw", (0, 0) + self.size, 0, (self.mode, 0, 1))]
         elif pfflags & DDPF_RGB:
             # Texture contains uncompressed RGB data
-            masks = {mask: ["R", "G", "B", "A"][i] for i, mask in enumerate(masks)}
-            rawmode = ""
-            if pfflags & DDPF_ALPHAPIXELS:
-                rawmode += masks[0xFF000000]
-            else:
+            if not pfflags & DDPF_ALPHAPIXELS:
                 self._mode = "RGB"
-            rawmode += masks[0xFF0000] + masks[0xFF00] + masks[0xFF]
 
-            self.tile = [("raw", (0, 0) + self.size, 0, (rawmode[::-1], 0, 1))]
+            # Each channel is derived using a mask of the pixel values
+            masks = {
+                mask: "RGB"[i]
+                for i, mask in enumerate(struct.unpack("<3I", header.read(12)))
```

Comment:
The previous version always read 16 bytes for masks – this will now read either 12 or 16 bytes. That seems odd?

Reply:
It isn't always necessary to read the last 4 bytes, and therefore seems more efficient to not do so where possible.

If you are concerned about the file pointer position, that's not relevant, as it is just reading from
https://github.com/python-pillow/Pillow/blob/697c24bc7dee8a679ae18d6267cc333f79a32dca/src/PIL/DdsImagePlugin.py#L127

Reply:
I think it would actually be more efficient and clearer to _not_ do another read-and-struct-unpack? (Would also mean that the entire [`DDS_PIXELFORMAT`](https://learn.microsoft.com/en-us/windows/win32/direct3ddds/dds-pixelformat) is always consumed.)